File: src/tools/connect.ts
import type { Connector } from "../connectors/base.js";
import type { SourceConfig, TableInfo } from "../types.js";
import { ExcelConnector } from "../connectors/excel.js";
import { CsvConnector } from "../connectors/csv.js";
import { MySQLConnector } from "../connectors/mysql.js";
import { CatalogStore } from "../catalog/store.js";
import { QueryEngine } from "../query/engine.js";
import { QueryRouter } from "../query/router.js";
import { collectSchema } from "../catalog/schema.js";

/**
 * Global connector registry — shared across tools.
 */
const connectors = new Map<string, Connector>();

export function getConnectors(): Map<string, Connector> {
  return connectors;
}

export function getConnector(sourceId: string): Connector | undefined {
  return connectors.get(sourceId);
}

/**
 * connect_source tool handler.
 */
export async function handleConnectSource(
  params: Record<string, unknown>,
  catalog: CatalogStore,
  engine?: QueryEngine,
  router?: QueryRouter,
): Promise<{ sourceId: string; tables: TableInfo[] }> {
  const config = params as unknown as SourceConfig;

  let connector: Connector;
  switch (config.type) {
    case "excel":
      connector = new ExcelConnector();
      break;
    case "csv":
      connector = new CsvConnector();
      break;
    case "mysql":
      connector = new MySQLConnector();
      break;
    default:
      throw new Error(`Unsupported source type: ${(config as { type: string }).type}`);
  }

  await connector.connect(params);

  const sourceId = connector.sourceId;
  connectors.set(sourceId, connector);

  // Persist source to catalog
  catalog.upsertSource(sourceId, connector.sourceType, params);

  // Collect schema
  const tables = await collectSchema(connector, catalog);

  // Register tables into the shared query engine DuckDB (for Excel/CSV)
  if (engine) {
    await connector.registerToDuckDB(engine.getDatabase());
  }

  // Register connector in query router for routing decisions
  if (router) {
    router.registerConnector(sourceId, connector, tables.map((t) => t.name));
  }

  return { sourceId, tables };
}

/**
 * list_tables tool handler.
 */
export async function handleListTables(
  params: Record<string, unknown>,
  catalog: CatalogStore,
): Promise<
  Array<{
    source_id: string;
    table_name: string;
    row_count: number;
    column_count: number;
    semantic_status: string;
  }>
> {
  const sourceId = params.sourceId as string | undefined;
  return catalog.getTables(sourceId);
}

File: lucid_skill/tools/connect.py
"""Global connector registry + connect/list handlers."""

from __future__ import annotations

from typing import Any

from lucid_skill.catalog.schema import collect_schema
from lucid_skill.catalog.store import CatalogStore
from lucid_skill.connectors.base import Connector
from lucid_skill.query.engine import QueryEngine
from lucid_skill.query.router import QueryRouter
from lucid_skill.types import TableInfo

_connectors: dict[str, Connector] = {}


def get_connectors() -> dict[str, Connector]:
    """Return the global connector registry."""
    return _connectors


def get_connector(source_id: str) -> Connector | None:
    """Return a connector by source ID, or None if not found."""
    return _connectors.get(source_id)


def handle_connect_source(
    params: dict[str, Any],
    catalog: CatalogStore,
    engine: QueryEngine | None = None,
    router: QueryRouter | None = None,
) -> dict[str, Any]:
    """Connect a data source. Returns {source_id, tables}."""
    from lucid_skill.connectors.csv_conn import CsvConnector
    from lucid_skill.connectors.excel_conn import ExcelConnector

    src_type = params.get("type")

    connector_map: dict[str, type] = {
        "csv": CsvConnector,
        "excel": ExcelConnector,
    }

    try:
        from lucid_skill.connectors.mysql_conn import MySQLConnector

        connector_map["mysql"] = MySQLConnector
    except ImportError:
        pass

    try:
        from lucid_skill.connectors.postgres_conn import PostgresConnector

        connector_map["postgresql"] = PostgresConnector
    except ImportError:
        pass

    cls = connector_map.get(src_type)
    if not cls:
        raise ValueError(f"Unsupported source type: {src_type}")

    connector: Connector = cls()
    connector.connect(params)

    source_id = connector.source_id
    _connectors[source_id] = connector

    # Persist source to catalog
    catalog.upsert_source(source_id, connector.source_type, params)

    # Collect schema
    tables: list[TableInfo] = collect_schema(connector, catalog)

    # Register tables into the shared query engine DuckDB (for Excel/CSV)
    if engine:
        connector.register_to_duckdb(engine.get_database())

    # Register connector in query router for routing decisions
    if router:
        router.register_connector(source_id, connector, [t.name for t in tables])

    return {"source_id": source_id, "tables": tables}


def handle_list_tables(
    params: dict[str, Any],
    catalog: CatalogStore,
) -> list[dict[str, Any]]:
    """List tables, optionally filtered by sourceId."""
    source_id = params.get("sourceId") or params.get("source_id")
    return catalog.get_tables(source_id)
